evm_arithmetization/src/cpu/kernel/asm/core/exception.asm

```diff
@@ -165,6 +167,88 @@ global exc_stack_overflow_check_stack_length:
     %jumpi(fault_exception)
     PANIC
 
+global exc_stop:
+    // Here, we need to check that the final registers have the correct value.
+    // stack: trap_info
+    PUSH @SEGMENT_REGISTERS_STATES
+    %add_const(6)
+    // Is the current stack_len is 2, then the stack was empty before the exception and there's no stack top.
+    // stack: addr_registers, trap_info
+    PUSH 3 
+    %stack_length
+    SUB
+    // First, check the stack length.
+    // stack: stack_len-2 = stack_len_before_exc, addr_registers, trap_info
+    DUP2 %add_const(2)
+    MLOAD_GENERAL
+    // stack: stored_stack_length, stack_len_before_exc, addr_registers, trap_info
+    DUP2 %assert_eq
+
+    // Now, check that we end up with the correct stack_top.
+    // stack: stack_len_before_exc, addr_registers, trap_info
+    DUP1 PUSH 0 LT
+    // stack: 0 < stack_len_before_exc, stack_len_before_exc, addr_registers, trap_info
+    PUSH 1 DUP3 SUB
+    // stack: stack_len_before_exc - 1, 0 < stack_len_before_exc, stack_len_before_exc, addr_registers, trap_info
+    MUL
+    // If the previous stack length is 0, we load the first value in the stack segment:
+    // we do not need to constrain the value in that case, so this is just to avoid a jumpi.
+    // Not having a jumpi provides a constant number of operations, which is better for segmentation.
+    // stack: (stack_len_before_exc - 1) * (0 < stack_len_before_exc), stack_len_before_exc, addr_registers, trap_info
+    PUSH @SEGMENT_STACK
+    GET_CONTEXT
+    %build_address
+    // stack: stack_top_before_exc_addr, stack_len_before_exc, addr_registers, trap_info
+    MLOAD_GENERAL
+    // stack: stack_top_before_exc, stack_len_before_exc, addr_registers, trap_info
+    DUP3 %add_const(3)
+    MLOAD_GENERAL
+    // stack: stored_stack_top, stack_top_before_exc, stack_len_before_exc, addr_registers, trap_info
+    SUB MUL
+    // stack: (stored_stack_top - stack_top_before_exc) * stack_len_before_exc, addr_registers, trap_info
+    %assert_zero
+
+    // Check the program counter.
+    // stack: addr_registers, trap_info
+    DUP2 PUSH 0xFFFFFFFF AND
+    // stack: program_counter, addr_registers, trap_info
+    DUP2
+    MLOAD_GENERAL
+    // stack: public_pc, program_counter, addr_registers, trap_info
+    %assert_eq
+
+    // Check is_kernel_mode.
+    // stack: addr_registers, trap_info
+    DUP2 %shr_const(32)
+    PUSH 0xFFFFFFFF AND
+    // stack: is_kernel_mode, addr_registers, trap_info
+    DUP2 %increment
+    MLOAD_GENERAL
+    %assert_eq
+
+    // Check the gas used.
+    // stack: addr_registers, trap_info
+    SWAP1 %shr_const(192)
+    PUSH 0xFFFFFFFF AND
+    // stack: gas_used, addr_registers
+    DUP2 %add_const(5)
+    MLOAD_GENERAL
+    %assert_eq
+
+    // Check the context.
+    // stack: addr_registers
+    %add_const(4)
+    MLOAD_GENERAL
+    // stack: stored_context
+    GET_CONTEXT
+    %assert_eq
+    // stack: (empty)
+    PUSH 1
+    ISZERO
```

Comment:
Hmm this would need to be explicitely documented here, with a warning to not touch it in case someone was to refactor the Kernel for efficiency at some point

Reply:
Is it clearer now?